internal/pkg/v1/action/catalog_delete.go

```diff
@@ -0,0 +1,69 @@
+package action
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	olmv1catalogd "github.com/operator-framework/catalogd/api/v1"
+
+	"github.com/operator-framework/kubectl-operator/pkg/action"
+)
+
+type CatalogDelete struct {
+	config      *action.Configuration
+	CatalogName string
+	DeleteAll   bool
+
+	Logf func(string, ...interface{})
+}
+
+func NewCatalogDelete(cfg *action.Configuration) *CatalogDelete {
+	return &CatalogDelete{
+		config: cfg,
+		Logf:   func(string, ...interface{}) {},
+	}
+}
+
+func (cd *CatalogDelete) Run(ctx context.Context) ([]string, error) {
+	// validate
+	if cd.DeleteAll && cd.CatalogName != "" {
+		return nil, errNameAndSelector
+	}
+
+	// delete single, specified catalog
+	if !cd.DeleteAll {
+		return nil, cd.deleteCatalog(ctx, cd.CatalogName)
+	}
+
+	// delete all existing catalogs
+	var catsList olmv1catalogd.ClusterCatalogList
+	if err := cd.config.Client.List(ctx, &catsList); err != nil {
+		return nil, err
+	}
+	if len(catsList.Items) == 0 {
+		return nil, errNoResourcesFound
+	}
+
+	errs := make([]error, 0, len(catsList.Items))
+	names := make([]string, 0, len(catsList.Items))
+	for _, cat := range catsList.Items {
+		names = append(names, cat.Name)
+		if err := cd.deleteCatalog(ctx, cat.Name); err != nil {
+			errs = append(errs, fmt.Errorf("failed deleting catalog %q: %w", cat.Name, err))
+		}
+	}
```

Comment:
I think you can avoid listing separately with the controllerruntime client's `DeleteAllOf`.

Reply:
I specifically avoided using `DeleteAllOf` and went with list -> delete, so that we:
* know what catalogs we are going to delete
* can wait for each one separately
* can return a more detailed/helpful message back to the user (either which catalogs were deleted or which failed deletion and why each of those failed).

Reply:
That's a good point, it's certainly helpful to have detailed error messages for failing to delete specific catalogs. 

I am a little worried about sending too many API requests on mass deletes - but hopefully mass deleting catalogs shouldn't be a common operation. `DeleteAllOf` also comes with support for label selectors which we may need to support in future, though we can get pretty much the same thing with the initial List.

Reply:
I agree, this is a concern. We probably might also at some point consider what level of performance are we expecting from the CLI.

Specifically regarding `DeleteAllOf`, if separate deletes turn out to be problematic, we can probably make a tradeoff between being performant and being specific in output to users. As a middle way, we could still leave the initial `List()` call and just combine it with `DeleteAllOf` like you suggested.
I'll look into the behavior more in depth and if we won't be missing much then potentially adjust it here or in a later PR.